src/main/kotlin/uk/gov/justice/digital/hmpps/prisonerfromnomismigration/visits/MigrationVisitsMessageListener.kt

```diff
@@ -1,60 +1,48 @@
-import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visits.VisitMessages.CANCEL_MIGRATE_VISITS
-import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visits.VisitMessages.MIGRATE_VISIT
-import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visits.VisitMessages.MIGRATE_VISITS
-import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visits.VisitMessages.MIGRATE_VISITS_BY_PAGE
-import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visits.VisitMessages.MIGRATE_VISITS_STATUS_CHECK
-import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.visits.VisitMessages.RETRY_VISIT_MAPPING
+import uk.gov.justice.digital.hmpps.prisonerfromnomismigration.service.VisitId
 import java.util.concurrent.CompletableFuture
 
 @Service
 class MigrationVisitsMessageListener(
-  private val objectMapper: ObjectMapper,
-  private val visitsMigrationService: VisitsMigrationService,
+  objectMapper: ObjectMapper,
+  visitsMigrationService: VisitsMigrationService,
+) : MigrationMessageListener<VisitsMigrationFilter, VisitId, NomisVisit, VisitNomisMapping>(
+  objectMapper,
+  visitsMigrationService,
 ) {
 
-  private companion object {
-    val log: Logger = LoggerFactory.getLogger(this::class.java)
-  }
-
   @SqsListener(VISITS_QUEUE_ID, factory = "hmppsQueueContainerFactoryProxy")
   @WithSpan(value = "dps-syscon-migration_visits_queue", kind = SpanKind.SERVER)
-  fun onMessage(message: String, rawMessage: Message): CompletableFuture<Void>? {
-    log.debug("Received message {}", message)
-    val migrationMessage: MigrationMessage<VisitMessages, *> = message.fromJson()
-    return CoroutineScope(Dispatchers.Default).future {
-      runCatching {
-        when (migrationMessage.type) {
-          MIGRATE_VISITS -> visitsMigrationService.divideVisitsByPage(migrationContext(message.fromJson()))
-          MIGRATE_VISITS_BY_PAGE -> visitsMigrationService.migrateVisitsForPage(migrationContext(message.fromJson()))
-          MIGRATE_VISIT -> visitsMigrationService.migrateVisit(migrationContext(message.fromJson()))
-          RETRY_VISIT_MAPPING -> visitsMigrationService.retryCreateVisitMapping(migrationContext(message.fromJson()))
-          MIGRATE_VISITS_STATUS_CHECK -> visitsMigrationService.migrateVisitsStatusCheck(migrationContext(message.fromJson()))
-          CANCEL_MIGRATE_VISITS -> visitsMigrationService.cancelMigrateVisitsStatusCheck(migrationContext(message.fromJson()))
-        }
-      }.onFailure {
-        log.error("MessageID:${rawMessage.messageId()}", it)
-        throw it
-      }
-    }.thenAccept { }
+  fun onVisitMessage(message: String, rawMessage: Message): CompletableFuture<Void>? {
+    return onMessage(message, rawMessage)
   }
 
-  private inline fun <reified T> String.fromJson(): T =
-    objectMapper.readValue(this, object : TypeReference<T>() {})
+  override fun parseContextFilter(json: String): MigrationMessage<*, VisitsMigrationFilter> {
+    return objectMapper.readValue(json)
+  }
```

Comment:
I don't understand star projections very well... but wouldn't you know the type at this point?

Reply:
No, because it is a class so the type gets lost at runtime unlike the reified inline functions